### PR TITLE
API request preview panel

### PR DESCRIPTION
**Pull request by deanhannigan**
## Description

Added a new debug panel into the the response section of the builder API query UI. The `Request` panel shows the resolved `url`, `body`, `headers`, `method` and `params` actually used in an outgoing request. 

Once the request has been made, the request panel will enable with redactions in place.
- All `environment` variables are always shown as placeholders as they are never exposed in the builder.
- All resolved auth data is redacted but still marked by type used.
- Some select fields like `x-api-key` are marked as redacted. Might be overkill

#### Bug fix
- Static parameters in datasource connections were resolving bindings. This isn't supported behaviour and must have been an oversight. Bindings are now properly ignored.

## Addresses
- https://github.com/Budibase/budibase/issues/19123

## Screenshots
recording

## Launchcontrol
Added request preview panel to the API query UX.

---
## Summary by cubic
Adds a Request preview panel to the API query builder with a toggle to switch between Request and Response. It shows the exact URL, method, headers, params, and body for previews—safely redacted—and is never exposed on execute.

- **New Features**
  - Sidebar `Switcher` toggles Request/Response; Request is disabled until a preview exists and selection resets to Response after send.
  - Request panel shows `QueryStats` and a JSON preview with secret placeholders rendered as badges using `secretTagPlugin`; env bindings in query and connection config are shown symbolically. Response now uses `QueryStats` too.
  - Server includes `extra.request` only for previews via `RestRequestPreview`/`RestPreviewConfig` and `IntegrationRequestOpts`; headers are normalized and redacted (preserving Basic, Bearer, OAuth2) with sensitive headers masked case-insensitively (e.g. `authorization`, `cookie`, `x-api-key`, `api-key`), params merge datasource defaults, and body is parsed (JSON, urlencoded, form-data) with form-data content-type included. OAuth2 401 retries retain the preview. For REST templates, the preview respects a swapped base URL and resolves static variables in the preview config base URL before swapping.

- **Bug Fixes**
  - Datasource `staticVariables` never apply bindings; any static value containing a binding is ignored.
  - Saved queries cannot surface a request preview on execute, even if they request it.

<sup>Written for commit 1a5f1839919a859cb05feba343045486f40b07e5. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/Budibase/budibase/pull/19443?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/server/src/api/controllers/query/index.ts [hunk 1]
```diff
@@ -262,9 +263,16 @@ function enrichParameters(
...
-    ctx.request.body.datasourceId
-  )
+  const rawDatasource = await sdk.datasources.get(ctx.request.body.datasourceId)
+  const { datasource, envVars } =
+    await sdk.datasources.enrichDatasourceWithValues(rawDatasource)
+  // Kept unresolved so the request preview can show bindings rather than the
+  // values they resolve to
+  const previewConfig: RestPreviewConfig = {
+    url: rawDatasource.config?.url,
+    defaultHeaders: rawDatasource.config?.defaultHeaders,
```
[the remark this answers is not in the document]
Comment (deanhannigan) [reply packages/server/src/api/controllers/query/index.ts:273]: The preview endpoint is builder-only, and builders can already read the datasource config

File: packages/server/src/integrations/rest.ts [hunk 2]
```diff
@@ -685,66 +718,201 @@ export class RestIntegration implements IntegrationBase {
...
+    built: BuiltRequest,
+    opts: IntegrationRequestOpts
+  ): RestRequestPreview | undefined {
+    const { previewFields, previewConfig } = opts
+    if (!previewFields) {
+      return undefined
+    }
+    try {
+      const fields: RestQuery = {
+        ...previewFields,
```
[the remark this answers is not in the document]
Comment (deanhannigan) [reply packages/server/src/integrations/rest.ts:847]: Pagination is never used in the preview

File: packages/server/src/api/controllers/query/index.ts [hunk 3]
```diff
@@ -392,6 +400,8 @@ export async function preview(
     nullDefaultSupport: query.nullDefaultSupport,
     queryId,
     datasource,
+    includeRequest: true,
```
[the remark this answers is not in the document]
Comment (deanhannigan) [reply packages/server/src/api/controllers/query/index.ts:403]: This is fixed

File: packages/builder/src/components/integration/APIEndpointViewer.svelte [hunk 4]
```diff
@@ -437,6 +441,7 @@
     existingQueryUnchanged ||
     newQueryIncomplete ||
     !isValidCustomUrl
+  $: requestPreview = response?.extra?.request as RestRequestPreview | undefined
```
[the remark this answers is not in the document]
Comment (deanhannigan) [reply packages/builder/src/components/integration/APIEndpointViewer.svelte:444]: This isnt an issue. The request/response panel shows stateful data relating to the last run of the query. It doesn't matter if you change the endpoint or connection. It will update on the next run to relay the new request state

File: packages/builder/src/components/integration/APIEndpointViewer.svelte [hunk 5]
```diff
@@ -728,6 +733,8 @@
 
       const result = await runQuery(builtQuery, schema)
       response = result.response
+      // A send always lands on the response, the request stays a click away
+      panelMode = "response"
```
Comment (melohagan) [inline packages/builder/src/components/integration/APIEndpointViewer.svelte:737]: Good shout 👌

File: packages/builder/src/components/integration/QueryStats.svelte [hunk 6]
```diff
@@ -0,0 +1,38 @@
+<script lang="ts">
+  import { Label } from "@budibase/bbui"
+
+  interface Props {
+    info?: { code: number; time: string; size: string }
+  }
+
+  let { info = undefined }: Props = $props()
+
+  const success = $derived(!!info && info.code >= 200 && info.code < 400)
```
Comment (melohagan) [inline packages/builder/src/components/integration/QueryStats.svelte:10]: NAB: Are 3xx really a success? Could maybe have an orange colour, but that's an edge case maybe not worth considering.

File: packages/server/src/api/routes/tests/queries/rest.spec.ts [hunk 7]
```diff
@@ -1650,7 +1733,7 @@ describe("rest", () => {
       }
     })
 
-    it("should resolve env var bindings in datasource staticVariables", async () => {
+    it("should never apply a static variable containing a binding", async () => {
```
Comment (melohagan) [inline packages/server/src/api/routes/tests/queries/rest.spec.ts:1736]: NAB: Could this technically be a breaking change?

File: packages/server/src/integrations/tests/rest.spec.ts [hunk 8]
```diff
@@ -726,6 +728,248 @@ describe("REST Integration", () => {
...
+
+      it("omits disabled headers", async () => {
+        integration = new RestIntegration({
+          url: "https://example.com",
+          defaultHeaders: { "x-connection": "from-connection" },
+        })
+        queueJsonResponse(() => {}, { foo: "bar" })
+
+        const fields = {
+          headers: { "x-query": "from-query" },
```
Comment (melohagan) [inline packages/server/src/integrations/tests/rest.spec.ts]: NAB: Should you also include a header that's not disabled and assert that it is present?

File: packages/server/src/integrations/tests/restUtils.spec.ts [hunk 9]
```diff
@@ -51,3 +57,135 @@ describe("getAttachmentHeaders", () => {
...
+      authType: RestAuthType.OAUTH2,
+    })
+    expect(sanitised.Authorization).toEqual(`Bearer ${SecretTag.OAUTH2}`)
+  })
+
+  it("should replace an api key held in a user named header", () => {
+    const sanitised = sanitiseHeaders({
+      headers: { "x-custom-key": "supersecret" },
+      authHeaderKeys: ["x-custom-key"],
+      authType: "apiKey",
```
Comment (melohagan) [inline packages/server/src/integrations/tests/restUtils.spec.ts]: How do I replicate this test manually through the UI?

File: packages/server/src/integrations/tests/restUtils.spec.ts [hunk 10]
```diff
@@ -51,3 +57,135 @@ describe("getAttachmentHeaders", () => {
...
+    })
+    expect(sanitised["x-custom-key"]).toEqual(SecretTag.API_KEY)
+  })
+
+  it("should replace sensitive headers which did not come from an auth config", () => {
```
Comment (melohagan) [inline packages/server/src/integrations/tests/restUtils.spec.ts:89]: Nice test 🧑‍🍳

File: packages/server/src/integrations/utils/restUtils.ts [hunk 11]
```diff
@@ -1,5 +1,111 @@
...
 import type { Headers as NodeFetchHeaders } from "node-fetch"
+import { JSONValue, RestAuthType, SecretTag } from "@budibase/types"
+import { findHBSBlocks } from "@budibase/string-templates"
+
+// Sensitive Header
+const SENSITIVE_HEADERS = [
+  "authorization",
+  "proxy-authorization",
+  "cookie",
+  "x-api-key",
```
Comment (melohagan) [inline packages/server/src/integrations/utils/restUtils.ts:14]: NAB: The authorization headers definitely makes sense as sensitive, however `x-api-key` and `api-key` are technically unofficial, although they are used quite frequently. 

Still it is not a catch all, apparently `X-Goog-Api-Key` is an actual header used by Google for example: https://docs.cloud.google.com/docs/authentication/api-keys-use

I guess you could do a case-insensitive contains on `api-key` or would be better to have all headers `Redacted` by default, with the option to click-to-reveal?

File: packages/server/src/integrations/utils/restUtils.ts [hunk 12]
```diff
@@ -1,5 +1,111 @@
...
+    default:
+      return SecretTag.GENERIC
+  }
+}
+
+// Retains an auth scheme prefix such as "Bearer" - it aids debugging and is
+// not itself sensitive - while replacing the credential which follows it.
+function redactValue(value: string, tag: string): string {
+  const [scheme, ...rest] = value.split(" ")
+  if (rest.length && /^(basic|bearer)$/i.test(scheme)) {
```
Comment (melohagan) [inline packages/server/src/integrations/utils/restUtils.ts:35]: NAB: This is missing `Authorization: ApiKey <key>`

File: packages/server/src/integrations/utils/restUtils.ts [hunk 13]
```diff
@@ -1,5 +1,111 @@
...
+  headers: unknown
+  authHeaderKeys?: string[]
+  authType?: string
+}): Record<string, string> {
+  const authKeys = new Set(authHeaderKeys.map(key => key.toLowerCase()))
+  const sanitised: Record<string, string> = {}
+  for (const [key, rawValue] of Object.entries(normaliseHeaders(headers))) {
+    const value = String(rawValue)
+    const lowerKey = key.toLowerCase()
+    const tag = authKeys.has(lowerKey)
```
Comment (melohagan) [inline packages/server/src/integrations/utils/restUtils.ts]: NAB: Avoid nested ternaries.

File: packages/server/src/integrations/rest.ts [hunk 14]
```diff
@@ -153,12 +162,24 @@ interface ParsedResponse {
...
 
+interface BuiltRequest {
+  url: string
+  input: RequestInit
+  authHeaders: Record<string, string>
+  authType?: string
+}
+
+interface RequestOpts extends IntegrationRequestOpts {
+  retry401?: boolean
```
Comment (melohagan) [inline packages/server/src/integrations/rest.ts:180]: Why would you ever retry a 401?

File: packages/server/src/integrations/rest.ts [hunk 15]
```diff
@@ -685,66 +718,201 @@ export class RestIntegration implements IntegrationBase {
...
+    }
+  }
+
+  // Headers derived from the body, such as the multipart content-type
+  private resolveHeaders(built: BuiltRequest): Record<string, string> {
+    const headers = normaliseHeaders(built.input.headers)
+    try {
+      const request = new Request(built.url, built.input)
+      for (const [key, value] of request.headers.entries()) {
+        if (!Object.keys(headers).some(k => k.toLowerCase() === key)) {
```
Comment (melohagan) [inline packages/server/src/integrations/rest.ts:822]: Do you need to lowercase the key as well?

File: packages/server/src/threads/query.ts [hunk 16]
```diff
@@ -369,6 +396,63 @@ class QueryRunner {
     return value
   }
 
+  // Builds the display-oriented preview
+  async buildPreview(
+    fields: Record<string, any>,
+    parameters: Record<string, any>
```
Comment (melohagan) [inline packages/server/src/threads/query.ts]: Avoid using **any** if possible.

File: packages/types/src/sdk/datasources.ts [hunk 17]
```diff
@@ -171,12 +171,26 @@ export type ConnectionInfo = {
...
+// will use
+export interface RestPreviewConfig {
+  url?: string
+  defaultHeaders?: Record<string, string>
+  defaultQueryParameters?: Record<string, string>
+}
+
+export interface IntegrationRequestOpts {
+  includeRequest?: boolean
+  previewFields?: Record<string, any>
```
Comment (melohagan) [inline packages/types/src/sdk/datasources.ts]: Avoid **any** type if possible (even if its just to use **unknown**)

File: packages/types/src/sdk/datasources.ts [hunk 18]
```diff
@@ -171,12 +171,26 @@ export type ConnectionInfo = {
...
+  previewConfig?: RestPreviewConfig
+}
+
 export interface IntegrationBase {
-  create?(query: any): Promise<any[] | any>
-  read?(query: any): Promise<any[] | any>
-  update?(query: any): Promise<any[] | any>
-  delete?(query: any): Promise<any[] | any>
-  patch?(query: any): Promise<any[] | any>
+  create?(query: any, opts?: IntegrationRequestOpts): Promise<any[] | any>
```
Comment (melohagan) [inline packages/types/src/sdk/datasources.ts:189]: NAB: Same here, update these **any** if its not too much hassle.

**Review by melohagan — CHANGES_REQUESTED** [review]
Pulled it down locally and the feature is great! Looks slick, and handy. Great work!

I've left mostly nitpicks, so should be good to go after a quick pass through.

File: packages/server/src/integrations/rest.ts (hunk 14 above)
Comment (deanhannigan) [reply packages/server/src/integrations/rest.ts:180]: https://github.com/Budibase/budibase/pull/15805

File: packages/server/src/api/routes/tests/queries/rest.spec.ts (hunk 7 above)
Comment (deanhannigan) [reply packages/server/src/api/routes/tests/queries/rest.spec.ts:1736]: No, static variables were never supposed to have bindings. We don't even provide a means to do this in the UI

File: packages/builder/src/components/integration/QueryStats.svelte (hunk 6 above)
Comment (deanhannigan) [reply packages/builder/src/components/integration/QueryStats.svelte:10]: 300 responses are just redirection statuses, so they are technically successful. An error in this case is considered just considered 400+

File: packages/server/src/integrations/utils/restUtils.ts (hunk 12 above)
Comment (deanhannigan) [reply packages/server/src/integrations/utils/restUtils.ts:35]: `ApiKey` isn't a core auth type that we support. `Basic` and `bearer` both get replaced with `Auth credential` to highlight that they came from the connection auth. `ApiKey` just gets replaced as `Redacted` as its in the sensitive header list.

Screenshot 2026-08-10 at 17 38 07

File: packages/server/src/integrations/rest.ts (hunk 14 above)
Comment (melohagan) [reply packages/server/src/integrations/rest.ts:180]: Nice!

File: packages/server/src/integrations/rest.ts (hunk 15 above)
Comment (deanhannigan) [reply packages/server/src/integrations/rest.ts:822]: It wouldn't hurt but `request.headers.entries()` produces lowercase values already

Comment (melohagan) [reply packages/server/src/integrations/rest.ts:822]: Ah nice one, that's grand then!

File: packages/server/src/integrations/tests/restUtils.spec.ts (hunk 9 above)
Comment (deanhannigan) [reply packages/server/src/integrations/tests/restUtils.spec.ts]: Niiiice catch. This should have been removed. We don't currently currently support this as a core connection auth type so it was completely dead 👍

File: packages/server/src/integrations/utils/restUtils.ts (hunk 11 above)
Comment (deanhannigan) [reply packages/server/src/integrations/utils/restUtils.ts:14]: I honestly threw the in `x-api-key` and `api-key` keys in as a nice-to-have. I think the case insensitive element could be a nice update, but we may end up chasing to accommodate headers outside our control.

We could possibly allow people to choose to redact in the UI before sending. Nice future enhancement maybe

Comment (melohagan) [reply packages/server/src/integrations/utils/restUtils.ts:14]: Or even just a toggle show/hide all headers, hidden by default. It's grand as is, but could be worth thinking about in future.

**Review by melohagan — APPROVED** [review]
LGTM!
